modularise namemanager

bimmlerd opened the pull request:
Modularisation of the FQDN NameManager. May be easier to review commit by commit, but it is a bit of code shuffling unfortunately.

The main pain point was to avoid the import cycle mentioned in the commit messages, which is what motivated moving it out of `fqdn` into its own package.

bimmlerd committed:
fqdn: move the NameManager into its own package

This is a preparation change to modularising the name manager into its
own cell. Pulling it out from the fqdn package will allow depending on
the endpointmanager package, which would have otherwise caused an import
cycle. The cycle would have been fqdn -> epmgr -> ep -> fqdn, but with
pulling the name manager out of fqdn proper, we can have it depend both
on fqdn and the ep packages.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

bimmlerd committed:
fqdn: make NameManager an interface

Another preparation commit to modularize the NameManager, though this
one could conceivably be squashed into the actual modularisation commit.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

bimmlerd committed:
fqdn: modularise NameManager

Pull the namemanager into its own cell and clean up a part of the
entanglement with the agent: get rid of the `GetEndpointDNSInfo`
callback to the agent.

There still is a fair bit of back and forth between the name manager and
the daemon, but for now that's considered part of the API of
NameManager. One consequence of this is that I can't just yet make the
NameManagerConfig a proper cell.Config - there's a few places (e.g. in
EP) where we access the MinTTL config without much access to the
hive. Refactoring that would increase the scope of this PR too much, so
we insert a gentle hack.

I've had to mock out an EndpointManager-like thing for the benchmark,
other than that there's very little new code.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

bimmlerd committed:
fqdn: move FQDN API handlers to namemanager

Now that we've modularised the name manager, we can also move the
related cilium API handlers out of the daemon/cmd/fqdn.

This also allows slimming down the NameManager API further, hiding
internals.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

dylandreimerink reviewed (approved):
Cell/Hive parts look good, can't really judge FQDN part.

bimmlerd committed:
fqdn: simplify endpoint handling in namemanager

Since we can now directly depend on the endpoint package, get rid of the
intermediate type. Additionally, inline the endpoint fetching logic
which was in getEndpointsDNSInfo, since only the API handlers actually
use an endpointID as selector.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

gandro commented on pkg/fqdn/cache.go:
```diff
@@ -801,7 +801,7 @@ type DNSZombieMappings struct {
 	deletes        map[netip.Addr]*DNSZombieMapping
 	lastCTGCUpdate time.Time
-	nextCTGCUpdate time.Time // estimated
+	NextCTGCUpdate time.Time // estimated
```
Is it actually save to read this without taking the DNSZombieMappings mutex? I know that you only exported the field and didn't change any of the behavior, but looks from a cursory glance like this field is written to by a separate GC go  routine [here](https://github.com/cilium/cilium/blob/81076cd7c781b285d78c9302a02ce90718327c52/pkg/maps/ctmap/gc/gc.go#L192) and read without any mutex by the name manager. Might be worth hiding it behind a simpler `GetNextCTGCUpdateTime` getter that actually takes the mutex when reading it.

gandro commented on pkg/fqdn/cache.go:
```diff
@@ -484,8 +484,8 @@ func (c *DNSCache) lookupIPByTime(now time.Time, ip netip.Addr) (names []string)
-// entryExistsLocked returns true if this (name, IP) pair is known to the cache.
-func (c *DNSCache) entryExistsLocked(name string, ip netip.Addr) bool {
+// EntryExistsLocked returns true if this (name, IP) pair is known to the cache.
+func (c *DNSCache) EntryExistsLocked(name string, ip netip.Addr) bool {
```
nit/not-your-code: This seems to be the only public `Locked` function for DNSCache and it has only one call-site. If we refactored the call site (`NameManager.maybeRemoveMetadata`) to perform a batch lookup in (e.g. by passing in a list of name/ip pairs and returned a map indicating which ones exist - or using some kind of interior iterator), we could get rid of the public `DNSCache.RWMutex` and make it private to `DNSCache`. All other call sites that take the `DNSCache` lock seem to be methods of `DNSCache` itself.

gandro reviewed (approved):
Excellent work! While reviewing the patch, I found a potential data race and further opportunity for reducing API surface, but they are not introduced by this PR and thus should not block the refactor. 

The fact that the refactor made this easy to spot proofs its value.
